packages/middleware-endpoint-discovery/src/resolveEndpointDiscoveryConfig.ts

```diff
@@ -0,0 +1,46 @@
+import { EndpointCache } from "@aws-sdk/endpoint-cache";
+import { Credentials, Provider } from "@aws-sdk/types";
+
+export interface EndpointDiscoveryInputConfig {}
+
+interface PreviouslyResolved {
+  isCustomEndpoint: boolean;
+  credentials: Provider<Credentials>;
+  /**
+   * The size of the client cache storing endpoints from endpoint discovery operations.
+   * Defaults to 1000.
+   */
+  endpointCacheSize?: number;
+  /**
+   * Whether to call operations with endpoints given by service dynamically.
+   * Setting this config to `true` will enable endpoint discovery for all applicable operations.
+   * Setting it to `false` will explicitly disable endpoint discovery even though operations that
+   * require endpoint discovery will presumably fail. Leaving it to undefined means SDK only do
+   * endpoint discovery when it's required. Defaults to `undefined`.
+   */
+  endpointDiscoveryEnabled?: boolean | undefined;
+  endpointDiscoveryEnabledProvider: Provider<boolean | undefined>;
+}
+
+export interface EndpointDiscoveryResolvedConfig {
+  isCustomEndpoint: boolean;
+  credentials: Provider<Credentials>;
+  endpointDiscoveryCommandCtor: new (comandConfig: any) => any;
+  endpointCache: EndpointCache;
+  endpointDiscoveryEnabled: Provider<boolean | undefined>;
+  isClientEndpointDiscoveryEnabled: boolean;
+}
```

Comment:
Documented each type in [`226726e` (#2369)](https://github.com/aws/aws-sdk-js-v3/pull/2369/commits/226726e49d81ee6c0bef1f1aeb4380e6ce267118).

Reply:
The type for `endpointDiscoveryCommandCtor` was made lose because of type mismatch error which resulted in [CI failure](https://us-west-2.console.aws.amazon.com/codesuite/codebuild/119327258862/projects/sdk-staging-test/build/sdk-staging-test%3A69c7d9e9-9c2a-48a5-98d0-bdcb24864df9/?region=us-west-2):

```console
DynamoDBClient.ts(440,63): error TS2345: Argument of type 'typeof DescribeEndpointsCommand' is not assignable to parameter of type 'new (comandConfig: any) => Command<any, any, any, any, any>'.
--
761 | Types of construct signatures are incompatible.
762 | Type 'new (input: DescribeEndpointsCommandInput) => DescribeEndpointsCommand' is not assignable to type 'new (comandConfig: any) => Command<any, any, any, any, any>'.
763 | Construct signature return types 'DescribeEndpointsCommand' and 'Command<any, any, any, any, any>' are incompatible.
764 | The types of 'middlewareStack.concat' are incompatible between these types.
765 | Type '<InputType extends DescribeEndpointsCommandInput, OutputType extends DescribeEndpointsCommandOutput>(from: MiddlewareStack<InputType, OutputType>) => MiddlewareStack<...>' is not assignable to type '<InputType extends any, OutputType extends any>(from: MiddlewareStack<InputType, OutputType>) => MiddlewareStack<InputType, OutputType>'.
```